bin/o80_robot_ball_logger.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python3
+
+"""
+Dump the output of a tennicam_client 
+frontend and of a o80_pam frontend into a file,
+with format for each line:
+ repr( ( (ball_id,ball_time_stamp,ball_position,ball_velocity),
+         (robot_time_stamp,robot_positions,robot_velocities) ) )
+with:
+- ball_id : int
+- ball_time_stamp: int (nanoseconds)
+- ball_position: 3d tuple
+- ball_velocity: 3d tuple
+- robot_time_stamp: int (nanoseconds)
+- robot_positions: 4d tuple
+- robot_velocities: 4d tuple
+Dumping is done at tennicam_client running frequency
+"""
+
+import pathlib
+import signal_handler
+import tennicam_client
+import o80_pam
+from lightargs import BrightArgs
+
+TENNICAM_CLIENT_DEFAULT_SEGMENT_ID = "tennicam_client"
+O80_PAM_DEFAULT_SEGMENT_ID = "real_robot"
+DEFAULT_SAVE_FOLDER = pathlib.Path("/tmp")
+
+
+def _unique_path(
+    directory: pathlib.Path, name_pattern="o80_robot_ball_{:03d}"
+) -> pathlib.Path:
+    """
+    returns the path to a file /directory/tennicam_{x} that does not exists yet
+    """
+
+    counter = 0
+    while True:
+        counter += 1
+        path = directory / name_pattern.format(counter)
+        if not path.exists():
+            return path
+
+
+def _run(tennicam_segment_id: str, o80_pam_segment_id: str, filepath: pathlib.Path):
+    """
+    Creates an tennicam_client frontend and a o80_pam frontend and
+    uses them to get all ball observations and pam robot observations;
+    and dumping them in a corresponding string in filepath.
+    """
+
+    ball_frontend = tennicam_client.FrontEnd(tennicam_segment_id)
+    robot_frontend = o80_pam.FrontEnd(o80_pam_segment_id)
+    iteration = ball_frontend.latest().get_iteration()
+
+    ball_getters = ("get_ball_id", "get_time_stamp", "get_position", "get_velocity")
+    robot_getters = ("get_time_stamp", "get_positions", "get_velocities")
+
+    with filepath.open(mode="w") as f:
+        try:
+            while not signal_handler.has_received_sigint():
```

Comment:
For my understanding, which part of the while loop sets the frequency here? Will `ball_frontend.read(iteration)` wait until `iteration` is reached?

Reply:
either this, or if the iteration has past, reads it from the history